poly/joint.go

```diff
@@ -0,0 +1,286 @@
+	if err == nil {
+		r.Dealers = append(r.Dealers, dealer)
+	}
+	return resp, err
+}
+
+// ProduceSharedSecret will generate the sharedsecret relative to this receiver
+// it will throw an error if something is wrong such as not enough Dealers received
+func (r *Receiver) ProduceSharedSecret() (*SharedSecret, error) {
+	if len(r.Dealers) < 1 {
+		return nil, errors.New("Receiver has 0 Dealers in its data.Can't produce SharedSecret.")
+	}
+	pub := new(PubPoly)
+	//pub.InitNull(r.info.Suite, r.info.T, r.Dealers[0].Promise.PubPoly().GetB())
+	pub.InitNull(r.info.Suite, r.info.T, r.info.Suite.Point().Base())
+	share := r.info.Suite.Secret()
+	goodShare := 0
+	for index, _ := range r.Dealers {
+		// Only need T shares
+		if goodShare >= r.info.T {
+			break
+		}
+		// Compute secret shares of the shared secret = sum of the respectives shares of peer i
+		// For peer i , s = SUM fj(i)
+		// TODO WARNING : THIS IS TEST IMPLEMENTATION
+		// In reality we should receive a NEW state struct from the dealer which is Certified so
+		// we can call RevealShare
+		// In testing we don't care about malicous yet so we just create one here
+		state := new(State).Init(*r.Dealers[index].Promise)
+		s, e := state.RevealShare(r.index, r.Key)
+		//s, e := r.Dealers[index].State.RevealShare(r.index, r.Key)
+		if e != nil {
+			//TODO error handling function not implemented right now. Only used for testing / comparison.
+			// We must be able to tell which share failed and to implement the broadcast of that error to others receiver
+			// so they reconstruct the private polynomial of the malicious dealer and set their share themself
+			return nil, errors.New(fmt.Sprintf("Receiver %d could not reveal its share from Dealer %d promise : %v", r.index, index, e))
+		}
+		share.Add(share, s)
+
+		// Compute shared public polynomial = SUM of indiviual public polynomials
+		pub.Add(pub, r.Dealers[index].Promise.PubPoly())
+
+		goodShare += 1
+	}
+
+	if goodShare < r.info.T {
+		return nil, errors.New("Not enough shares received by the Receiver to construct its own share of the shared secret")
+	}
+
+	if val := pub.Check(r.index, share); val == false {
+		return nil, errors.New("Receiver's secret share of the shared secret could not be checked against the shared polynomial")
+	}
+
+	return &SharedSecret{
+		Pub:   pub,
+		Share: &share,
+		Index: r.index,
+	}, nil
+}
+
+// MARSHALLING side
+
+// PolyInfo marshalling :
+// It is important to note that it does not Marshal / Unmarshal
+// the SUITE used. The suite is generated by default in constants.go
+// If you want to change to another suite, just change SUITE = ....
+func (p *PolyInfo) MarshalSize() int {
+	return util.IntSize(p.N) +
+		util.IntSize(p.R) +
+		util.IntSize(p.T)
+}
+
+func (p *PolyInfo) Equal(p2 PolyInfo) bool {
+	return p.N == p2.N && p.R == p2.R && p.T == p2.T && p.Suite.String() == p2.Suite.String()
+}
+
+func (p *PolyInfo) MarshalBinary() ([]byte, error) {
```

Comment:
Sorry for not being able to keep up or reply earlier yesterday or today, due to the Zurich trip and my first lecture this afternoon…

Please do NOT directly push changes to the master branch, even when they seem like trivial or obvious bug fixes.  Commit them to an alternate branch and open a pull request for discussion, peer review, and signed.  You can keep your own persistent working branch if you don’t like creating temporary branches each time, and if anyone else needs to use your fixes right away, they can ‘git checkout’ your branch or merge the changes from your branch into theirs.

In particular, what might seem like a trivial or obvious “bug fix” to someone might seem more subtle, problematic, or multifaceted to someone else. :)  For example, you decided unilaterally that the right semantics for what happens when the reader encounters a null pointer is that it should silently create an instance of the appropriate type.  It’s not obvious to me that those are necessarily the right or appropriate semantics.  

For example, these “auto-construct” semantics for pointers are inconsistent with the behavior for slices, which as you observed, require the slice to be created with the appropriate length before the Read so that Read will know how many elements to decode.  This basic binary encoding obviously cannot instantiate a nil slice on Read, because the encoding doesn’t transmit metadata on how long the slice should be.  (Unlike protobuf or another more sophisticated encoding, for example).  Thus for this encoding the only two feasible semantic behaviors for reading a nil slice are to fail/panic or to treat a nil slice like an empty slice.  I’m not entirely sure which is better - treating it like a nil slice might be slightly more convenient when that’s what the developer wants, but it also might risk leading to harder-to-find bugs when the developer actually intended to initialize the slice to a non-nil length but just forgot to initialize it.  If the “forgot to initialize it” case is more common, that might actually argue for a fail/panic to be the right semantic behavior, so the error will be caught more immediately.  I’m actually not sure which of these two behaviors the code currently exhibits - forgetting to test, think about, decide on, and document that corner case is actually probably a meta-bug that needs to be fixed one way or another.  But certainly the behavior isn’t and can’t be just “create a non-nil slice of the ‘right’ length” because the ‘right’ length is unknown by Read.

So back to the semantics for nil pointers: should the right semantics be to create an instance?  Maybe, but I’m not sure.  That seems necessarily inconsistent with slice handling, at least.  Why was a pointer type used there in the first place in the type (usually a struct) containing it, and not just a direct non-pointer instance of the pointed-to type?  The protobuf encoding/decoding code (dedis/protobuf) takes a pointer-type struct element as indicating an “optional” protobuf field - which may or may not be the right thing, but the point is it’s semantically meaningful: a pointer-to-X is different from X and there might thus be a reason or need for the encoding/decoding code to treat them differently.  The simple fixed-size binary format, unlike protobuf, doesn’t have metadata with which to indicate the presence or absence of a field, so that particular interpretation of the meaning of a pointer-type struct element doesn’t make sense.  What exactly does make sense?  I’m not sure, but I’m hesitant to say the right behavior is just to create an instance automatically.  If the developer “always” wanted an instance (i.e., a mandatory, non-optional field), then shouldn’t he have just used a non-pointer embedded instance of the type in the struct instead of a pointer type?  If the fixed-size binary encoding requires developers to instantiate slices before Read, why shouldn’t it also require developers to instantiate pointers before Read?

At any rate, again the meta-point is that these kinds of things need to be discussed - or at least given the opportunity for discussion if needed - by using branches and pull requests before immediately committing to the master branch.

Bryan

> On Sep 18, 2015, at 3:51 PM, nikkolasg <notifications@github.com> wrote:
> 
> In poly/joint.go <https://github.com/DeDiS/crypto/pull/44#discussion_r39856269>:
> 
> > +
> > +// PolyInfo marshalling :
> > +// It is important to note that it does not Marshal / Unmarshal
> > +// the SUITE used. The suite is generated by default in constants.go
> > +// If you want to change to another suite, just change SUITE = ....
> > +func (p *PolyInfo) MarshalSize() int {
> > +	return util.IntSize(p.N) +
> > +		util.IntSize(p.R) +
> > +		util.IntSize(p.T)
> > +}
> > +
> > +func (p *PolyInfo) Equal(p2 PolyInfo) bool {
> > +	return p.N == p2.N && p.R == p2.R && p.T == p2.T && p.Suite.String() == p2.Suite.String()
> > +}
> > +
> > +func (p *PolyInfo) MarshalBinary() ([]byte, error) {
> I've added one more piece of code that i find useful is the ability to read/write simple "int". The code in the master branch only allow to use specific fixed-length type such as int32 or uint64. In my structs and many others, we use the generic type "int". What I did is simply writing the value as an int64 and reading it as an int64. So the int value is read transparently to the user and sent across the wire as a int64. 
> Would you be interested if I push that to the master branch or not ? 
> As always there are some tradeoff, here we must create a new variable of type int64, but I don't think that will make things slower than using the reflect library at all ;)
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/DeDiS/crypto/pull/44/files#r39856269>.
>